internal/metrics/metrics.go

```diff
@@ -0,0 +1,114 @@
+package metrics
+
+
+import (
+	"time"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	ms = 0.001
+	s  = 1000 * ms
+	m  = 60 * s
+	kb = 1024
+	mb = 1024*kb
+)
+
+type metrics struct {
+	registry *prometheus.Registry
+	endpoint string
+
+	// Custom metics
+	request_durations *prometheus.HistogramVec
+	response_sizes    *prometheus.HistogramVec
+
+}
+
+func (m *metrics) GetEndpoint() string {
+	return m.endpoint
+}
+
+/** Create a new metric instance
+ *
+ * The metrics are applies to a new non-global prometheus registry
+ */
+func NewMetrics(endpoint string) *metrics {
+	registry := prometheus.NewRegistry()
+	metrics := &metrics{
+		registry: registry,
+		endpoint: strings.TrimLeft(endpoint, "/"),
+
+		request_durations: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "vdsslice_durations_histogram_seconds",
+			Help:    "VDSslice latency distributions.",
+			Buckets: []float64{100*ms, 500*ms, 1*s, 2*s, 5*s, 20*s, 1*m, 2*m},
+		}, []string{"method", "path", "status", "cachehit"}),
+
+		response_sizes: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "vdsslice_response_sizes_histogram_bytes",
+			Help:    "VDSslice response size distributions.",
+			Buckets: []float64{100*kb, 1*mb, 5*mb, 10*mb, 20*mb, 50*mb, 100*mb, 200*mb},
+		}, []string{"method", "path", "status"}),
+	}
+
+	registry.MustRegister(metrics.request_durations)
+	registry.MustRegister(metrics.response_sizes)
+
+	return metrics;
+}
+
+/** New gin middleware for writing prometheus metrics */
+func NewGinMiddleware(metrics *metrics) gin.HandlerFunc {
+	return func(ctx *gin.Context) {
+		// Don't polute metrics with request towards the /metric endpoint
+		// itself
+		path := strings.Split(ctx.Request.URL.Path, "?")[0]
+		if strings.TrimLeft(path, "/") == metrics.endpoint {
+			ctx.Next()
+			return
+		}
+		
+		start := time.Now()
+		ctx.Next()
+
+		go func() {
+			method := ctx.Request.Method
+			status := http.StatusText(ctx.Writer.Status())
```

Comment:
Do we want to present status as string or as number? Now it prints "OK", "Not found" etc. Do we maybe want 200/400/404/500 instead?